process/transaction/interceptedTransaction.go

```diff
@@ -199,51 +199,64 @@ func (inTx *InterceptedTransaction) CheckValidity() error {
 			return err
 		}
 
+		err = inTx.verifyIfRelayedTxV3(inTx.tx)
+		if err != nil {
+			return err
+		}
+
 		inTx.whiteListerVerifiedTxs.Add([][]byte{inTx.Hash()})
 	}
 
 	return nil
 }
 
-func isRelayedTx(funcName string) bool {
-	return core.RelayedTransaction == funcName || core.RelayedTransactionV2 == funcName
+func isRelayedTx(funcName string, innerTx *transaction.Transaction) bool {
+	return core.RelayedTransaction == funcName ||
+		core.RelayedTransactionV2 == funcName ||
+		innerTx != nil
 }
 
-func (inTx *InterceptedTransaction) verifyIfRelayedTxV2(tx *transaction.Transaction) error {
-	funcName, userTxArgs, err := inTx.argsParser.ParseCallData(string(tx.Data))
-	if err != nil {
-		return nil
-	}
-	if core.RelayedTransactionV2 != funcName {
+func (inTx *InterceptedTransaction) verifyIfRelayedTxV3(tx *transaction.Transaction) error {
+	if tx.InnerTransaction == nil {
 		return nil
 	}
 
-	userTx, err := createRelayedV2(tx, userTxArgs)
-	if err != nil {
-		return err
+	innerTx := tx.InnerTransaction
+	if !bytes.Equal(innerTx.SndAddr, tx.RcvAddr) {
+		return process.ErrRelayedTxV3BeneficiaryDoesNotMatchReceiver
+	}
+	if len(innerTx.RelayerAddr) == 0 {
```

Comment:
why do you need a relayerAddress in the innerTX ? it is not needed. Duplicate information. You have the relayerAddress from the tx.SndAddr

Reply:
added it for an extra check, to be the same with (parent tx).SndAddr, as discussed with @AdoAdoAdo in order to prevent sandwich attacks

Reply:
that's good. As the user can choose the relayer.

Reply:
But add the check

Reply:
added